graph/push_v2.go

```diff
@@ -0,0 +1,249 @@
+	if _, err := p.poolAdd("push", localName); err != nil {
+		return err
+	}
+	defer p.poolRemove("push", localName)
+
+	tags, err := p.getImageTags(p.localRepo, imagePushConfig.Tag)
+	if err != nil {
+		return fmt.Errorf("error getting tags for %s: %s", localName, err)
+	}
+	if len(tags) == 0 {
+		return fmt.Errorf("no tags to push for %s", localName)
+	}
+
+	for _, tag := range tags {
+		if err := p.pushV2Tag(repo, p.localRepo, tag, imagePushConfig.OutStream, sf); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *v2Pusher) pushV2Tag(repo distribution.Repository, localRepo Repository, tag string, out io.Writer, sf *streamformatter.StreamFormatter) error {
+	logrus.Debugf("Pushing repository: %s:%s", repo.Name(), tag)
+
+	layerId, exists := localRepo[tag]
+	if !exists {
+		return fmt.Errorf("tag does not exist: %s", tag)
+	}
+
+	// TODO(tiborvass): @dmcgowan isn't this supposed to be outside the tags loop?
+	layersSeen := make(map[string]bool)
+
+	layer, err := p.graph.Get(layerId)
+	if err != nil {
+		return err
+	}
+
+	m := &manifest.Manifest{
+		Versioned: manifest.Versioned{
+			SchemaVersion: 1,
+		},
+		Name:         repo.Name(),
+		Tag:          tag,
+		Architecture: layer.Architecture,
+		FSLayers:     make([]manifest.FSLayer, 0, 4),
+		History:      make([]manifest.History, 0, 4),
+	}
+
+	var metadata runconfig.Config
+	if layer != nil && layer.Config != nil {
+		metadata = *layer.Config
+	}
+
+	for ; layer != nil; layer, err = p.graph.GetParent(layer) {
+		if err != nil {
+			return err
+		}
+
+		if layersSeen[layer.ID] {
+			break
+		}
+
+		logrus.Debugf("Pushing layer: %s", layer.ID)
+
+		if layer.Config != nil && metadata.Image != layer.ID {
+			if err := runconfig.Merge(&metadata, layer.Config); err != nil {
+				return err
+			}
+		}
+
+		jsonData, err := p.graph.RawJSON(layer.ID)
+		if err != nil {
+			return fmt.Errorf("cannot retrieve the path for %s: %s", layer.ID, err)
+		}
+
+		// Push v1 compatibility object for compatibility registry
+		// Set jsonData and layer.ID
+
+		dgst, err := p.graph.GetDigest(layer.ID)
+		switch err {
+		case nil:
+			_, err := repo.Blobs(nil).Stat(nil, dgst)
+			switch err {
+			case nil:
+				out.Write(sf.FormatProgress(stringid.TruncateID(layer.ID), "Image already exists", nil))
+			case distribution.ErrBlobUnknown:
+				dgst = ""
+			default:
+				out.Write(sf.FormatProgress(stringid.TruncateID(layer.ID), "Image push failed", nil))
+				return err
+			}
+		case ErrDigestNotSet:
+			// nop
+		case digest.ErrDigestInvalidFormat, digest.ErrDigestUnsupported:
+			return fmt.Errorf("error getting image checksum: %v", err)
+		}
+
+		// if digest was empty or not saved, or if blob does not exist on the remote repository,
+		// then fetch it.
+		if len(dgst) == 0 {
+			if pushDigest, err := p.pushV2Image(repo.Blobs(nil), layer, sf, out); err != nil {
+				return err
+			} else if pushDigest != dgst {
```

Comment:
I'll revert to the logic with an `exists bool` variable and use that instead of resetting `dgst = ""` when upon `ErrBlobUnknown`. Thanks!